src/Utilities/TaskItem.cs

```diff
@@ -194,8 +194,20 @@ public ICollection MetadataNames
         {
             get
             {
-                var metadataNames = new List<string>(_metadata?.Keys ?? Array.Empty<string>());
-                metadataNames.AddRange(FileUtilities.ItemSpecModifiers.All);
+                var metadataNames = new List<string>(capacity: MetadataCount);
+
+                if (_metadata is not null)
+                {
+                    foreach (string key in _metadata.Keys)
+                    {
+                        metadataNames.Add(key);
+                    }
+                }
+
+                foreach (string name in FileUtilities.ItemSpecModifiers.All)
+                {
+                    metadataNames.Add(name);
+                }
```

Comment:
What is the advantage of this code compared to calling `AddRange`? `AddRange` appears to detect `ICollection<T>` and use its `CopyTo` to bulk-add all items so it may be even faster than the array-optimized foreach.
https://github.com/dotnet/runtime/blob/77165a096318b30bb380d8c47b00a1c02a541dba/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/List.cs#L738-L766

Reply:
Thanks, that's a great observation. 4e0ea49e5d5e7fe7c12a4fc01c0323715e678d0e introduces three usages of `AddRange`.